### ISSUE TITLE
predictable memory usage by reflector

### ISSUE THREAD

**Issue by zhouhaibing089** [kubernetes/kubernetes#73497]
**What would you like to be added**:

Predictable memory usage by reflector. I mean, the memory used by reflector does not have to be linear to the total number of objects in apiserver, since the list call can be paged, so I would expect the memory used should just be enough to handle a single page ideally.

**Why is this needed**:

Considering when a controller starts up, the reflector tries to list from apiserver, now if the initial data size is pretty huge(300+MB?), the memory used to decode objects could be huge, I can see that CoreDNS implements something that converts the object to a trimmed version with important fields only, but in my opinion, it does not help much during initialization phase, as the initial memory still has to be allocated, as a result, the application can consume more than 1G memory initially while it only needs 300MB(or less as what is done in CoreDNS) once the list is done. More generally, the question is: is there a possibility that we make sure the memory is bounded even during `list`phase? Maybe something like a shared memory for decoding(a single page at most?), and then trimmed it down for event handlers to consume?

**Comment by roycaihw** [comment]
cc @jpbetz

**Comment by yue9944882** [comment]
cc @adohe

**Comment by johnbelamaric** [comment]
/remove-lifecycle-stale
/remove-triage support
/kind bug

I think this is a significant issue that should be addressed. The initial surge in memory utilization means that you have to set the memory limit fairly high to avoid OOM, even if on a an ongoing basis the memory utilization is much much lower. We should be able to process the list without consuming so much memory.

@wojtek-t @lavalamp

**Comment by wojtek-t** [comment]
Pagination + not waiting for all pages to be downloaded before processing them should visibly help with that.
@jpbetz was working on that and there should be visible improvement in 1.15 for it.

**Comment by jpbetz** [comment]
Reflectors typically list from the watch which does not paginate.

**Comment by lavalamp** [comment]
/assign @jpbetz 

Joe is working on this if it isn't already finished. There is no reason reflectors can't start paginating.

**Comment by lavalamp** [comment]
OK let's try to make a sensible comment, haha.

Re: OP: the memory usage is high probably not because of the initial list, but because everything is typically cached. A heap trace or whatever shows the allocator of the memory, which will be reflector, but reflector doesn't (shouldn't) keep references to things.

If reflectors are not paginating the initial list, we should definitely turn that on.

**Comment by johnbelamaric** [comment]
We slim down the in-memory cache by using our own smaller objects that just
keep the data we need. See
https://github.com/coredns/coredns/tree/master/plugin/kubernetes/object

**Comment by zhouhaibing089** [comment]
Sometime ago, I read this:

> Go is slow to release memory back to the OS but fast to request it

That basically means the application needs to have a higher memory limits.

**Comment by wojtek-t** [comment]
> If reflectors are not paginating the initial list, we should definitely turn that on.

Reflectors are listing (by default) from watchcache, and watchcache doesn't support pagination.

But your second comment was actually an important one - I don't think pagination would visibly help, because we are storing all objects in memory anyway (unless you're doing something like @johnbelamaric mentioned for CoreDNS).

> Go is slow to release memory back to the OS but fast to request it

Go 1.12 is waaaaaay better in that (and it's used in releases 1.14+)

**Comment by lavalamp** [comment]
> Reflectors are listing (by default) from watchcache, and watchcache doesn't support pagination.

Good point about the watchcache. But, I still think we should at least experiment with paginated lists rather than lists that use the watchcache, have we?

**Comment by wojtek-t** [comment]
> Good point about the watchcache. But, I still think we should at least experiment with paginated lists rather than lists that use the watchcache, have we?

We did in the past, but it was at the beginning of pagination feature (like 2-ish years ago?)
I agree we should probably try again.
